### ISSUE TITLE
Blank screen for 4 minutes - show splash screen earlier?

### ISSUE THREAD

**Issue by OliverEvans96** [jupyterlab/jupyterlab#5463]
Hi all,

I'm currently using jupyterlab over the internet, with a fairly slow connection, roughly 50-150kB/s, to the remote server. Meanwhile, I'm working on a fairly large notebook, although it's only around 300kB.

When I open jupyterlab in my browser, I'm noticing one GET request to `/lab/static/vendors~main.035a8bbdf5a51f2df3e1.js`. It's 21.2MB and took 3.8min to download. That seems quite large - I'm not sure what it is.

The issue is that the splash screen doesn't appear until this request finishes, so my screen is blank for 4 minutes before there's any indication that I've actually navigated to the correct page. So perhaps the splash screen should display earlier, or some type of loading bar?

Thanks!
Oliver

**Comment by jasongrout** [comment]
Great question. We've talked about having an initial splash screen embedded in the html of the page (i.e., not part of the js code that is downloaded). I think that's an excellent idea.

21MB seems quite large for the vendors js bundle. Is that the stock bundle, or have you installed a bunch of plugins? The vendors bundle size I see in the stock jlab is more like 6MB.

**Comment by OliverEvans96** [comment]
Good point - I do have quite a few extensions.

```
jovyan@jupyterlab-5bd659cc4c-tnlnm:~$ jupyter labextension list
JupyterLab v0.34.12
Known labextensions:
   app dir: /opt/conda/share/jupyter/lab
        @jupyter-widgets/jupyterlab-manager v0.37.4  enabled  OK
        @jupyterlab/hub-extension v0.11.0  enabled  OK
        @jupyterlab/plotly-extension v0.18.0  enabled  OK
        @jupyterlab/toc v0.5.0  enabled  OK
        bqplot v0.4.2  enabled  OK
        ipyvolume v0.5.1  enabled  OK
        jupyter-leaflet v0.9.0  enabled  OK
        jupyter-threejs v1.1.0  enabled  OK
        jupyterlab-datawidgets v5.1.0  enabled  OK
        jupyterlab-scales v0.2.1  enabled  OK
        jupyterlab_bokeh v0.6.2  enabled  OK
        jupyterlab_vim v0.9.0  enabled  OK
        nbdime-jupyterlab v0.5.0  enabled  OK
```

**Comment by jasongrout** [comment]
We should have better documentation and encouragement for extensions to dynamically load the bulk of their js. For example, in ipywidgets, we load the actual widget code only when it's used. We should do the same for leaflet, etc. Then you wouldn't have that code pulled down in the initial load.

**Comment by madsi1m** [comment]
We have the same issue, i think this can be increased if the vendors-main was smaller, or split into many little files so the browser can download them quicker in parallel.

A splash screen before the js download could be good for our users as well as it shows something is happening, rather than the user thinking the site is down because it is blank.

**Comment by jasongrout** [comment]
> We have the same issue, i think this can be increased if the vendors-main was smaller, or split into many little files so the browser can download them quicker in parallel.

This is done in 3.0, especially when using the new prebuilt extensions. In particular, the main jlab core bundle is now about 2.0MB.



> A splash screen before the js download could be good for our users as well as it shows something is happening, rather than the user thinking the site is down because it is blank.

The technical challenge with displaying a splash screen on the initial page is that we don't know if the theme is light or dark, so we don't know which kind of splash screen to display (the splash screen changes its color depending on if it is fading to a light or dark theme). Finding out the initial theme color in time for initial page load would involve some special-cased code that assumes the theme manager is available, assumes the structure of the theme manager settings, etc. It'll take some thought to do this elegantly.

**Comment by caponetto** [comment]
Hi @jasongrout. I understand the challenges with the theme but, no matter the theme being used, currently there is still a blank screen being shown. Depending on the environment, the delay could give the impression that something is wrong. I wonder if it would be possible to show a simple loading indicator on the blank screen. At least, users would know that they need to wait while something is happening.

**Comment by jiridanek** [comment]
> currently there is still a blank screen being shown

blank _white_ screen, right?

**Comment by krassowski** [comment]
> blank white screen, right?

Yes

So few points:
- Initial JupyterLab bundle is not as big and thus time to load a themed spinner (respecting light/dark) is short
- The initial JupyterLab bundle can get smaller if we move more things to async loading, see https://github.com/jupyterlab/jupyterlab/issues/16135
- Third-party extensions can be a big drag on the initial load time; I recently cut 4MB out of jupyter-ai (~7s time, depending on your server and connection speed) in https://github.com/jupyterlab/jupyter-ai/pull/1225 and the same was applied to `jupyterlab-git` in https://github.com/jupyterlab/jupyterlab-git/pull/1393; I would suggest to:
  - check from which extensions the large bundles are coming from (either inspecting network requests or disabling one by one)
  - open issues for these extensions, suggesting to use async imports or cutting deps they don't need
- deferring extension load could help; there is `deferredExtensions` stanza maybe JupyterLab or extensions could use it more (currently it is not used in core)?
